src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,529 @@
+                {
+                    intersectingSegments.add(Pair.of(leftSegment, rightSegment));
+                }
+            }
+        }
+        return intersectingSegments;
+    }
+
+    /**
+     * Check the atlas object for boundary nodes at its end
+     *
+     * @param object
+     *            The atlas object to check
+     * @return {@code true} if the object ends on a boundary
+     */
+    private static boolean endsWithBoundaryNode(final AtlasObject object)
+    {
+        if (!(object instanceof LineItem))
+        {
+            return false;
+        }
+        final LineItem lineItem = (LineItem) object;
+        final Atlas atlas = object.getAtlas();
+        final Location last = lineItem.asPolyLine().last();
+        final Stream<Point> points = Iterables.asList(atlas.pointsAt(last)).stream();
+        final Stream<Node> nodes = Iterables.asList(atlas.nodesAt(last)).stream();
+        return Stream.concat(points, nodes).anyMatch(SyntheticBoundaryNodeTag::isBoundaryNode);
+    }
+
+    /**
+     * Default constructor
+     *
+     * @param configuration
+     *            the JSON configuration for this check
+     */
+    public WaterWayCheck(final Configuration configuration)
+    {
+        super(configuration);
+        this.elevationUtils = new ElevationUtilities(configuration);
+        this.waterwaySinkTagFilter = this.configurationValue(configuration,
+                "waterway.sink.tags.filters", WATERWAY_SINK_TAG_FILTER_DEFAULT,
+                TaggableFilter::forDefinition);
+        this.waterwayTagFilter = this.configurationValue(configuration, "waterway.tags.filters",
+                WATERWAY_TAG_FILTER_DEFAULT, TaggableFilter::forDefinition);
+
+        /* Ocean data */
+        this.validOceanTags = TaggableFilter.forDefinition(
+                this.configurationValue(configuration, "ocean.valid", DEFAULT_VALID_OCEAN_TAGS));
+        this.oceanBoundaryTags = TaggableFilter.forDefinition(this.configurationValue(configuration,
+                "ocean.boundary", DEFAULT_OCEAN_BOUNDARY_TAGS));
+
+        /* End ocean data */
+        /* Elevation settings */
+        this.minResolutionDistance = this.configurationValue(configuration,
+                "waterway.elevation.resolution.min.uphill", MIN_RESOLUTION_DISTANCE.asMeters(),
+                Distance::meters);
+        this.minDistanceStartEndElevationUphill = this.configurationValue(configuration,
+                "waterway.elevation.distance.min.start.end",
+                Distance.FIFTEEN_HUNDRED_FEET.asMeters(), Distance::meters);
+        /* End elevation settings */
+
+    }
+
+    /**
+     * Check if a line crosses an coastline line
+     *
+     * @param line
+     *            The line to check
+     * @return {@code true} if the line crosses a coastline
+     */
+    private boolean doesLineCrossCoast(final LineItem line)
+    {
+        final List<LineItem> lines = new ArrayList<>();
+        line.getAtlas()
```

Comment:
Same comment as https://github.com/osmlab/atlas-checks/pull/332/files#r485768561 on storing the atlas and passing it to helpers.

Reply:
I was kind of surprised by this. I expected that call to be pretty cheap. :(